backend/media/src/main/java/co/airy/core/media/MessageDataResolver.java

```diff
@@ -0,0 +1,116 @@
+
+import static co.airy.model.metadata.MetadataRepository.getId;
+import static co.airy.model.metadata.MetadataRepository.newMessageMetadata;
+
+@Component
+public class MessageDataResolver {
+    private final Logger log = AiryLoggerFactory.getLogger(MessageDataResolver.class);
+    private final String applicationCommunicationMetadata = new ApplicationCommunicationMetadata().name();
+    private final KafkaProducer<String, Metadata> producer;
+    private final MediaUpload mediaUpload;
+    private final ContentMapper mapper;
+    private final ExecutorService executor;
+
+    public MessageDataResolver(KafkaProducer<String, Metadata> producer,
+                               MediaUpload mediaUpload,
+                               ContentMapper mapper) {
+        this.producer = producer;
+        this.mediaUpload = mediaUpload;
+        this.mapper = mapper;
+        this.executor = Executors.newSingleThreadExecutor();
+    }
+
+    public void onMessageMediaRequest(String messageId, MessageMediaRequest messageMediaRequest) {
+        executor.submit(() -> processMessageMediaRequests(messageId, messageMediaRequest));
+    }
+
+    private void processMessageMediaRequests(String messageId, MessageMediaRequest messageMediaRequest) {
+        final Message message = messageMediaRequest.getMessage();
+        final Map<String, String> metadataMap = Optional.ofNullable(messageMediaRequest.getMetadata()).orElse(new HashMap<>());
+
+        final List<Content> contentList = mapper.renderWithDefaultAndLog(message, metadataMap);
+
+        for (Content content : contentList) {
+            if (!(content instanceof DataUrl)) {
+                continue;
+            }
+
+            final String sourceUrl = ((DataUrl) content).getUrl();
+
+            try {
+                final URL url = new URL(sourceUrl);
+
+                if (!mediaUpload.isUserStorageUrl(url) && !hasPersistentUrl(metadataMap, sourceUrl)) {
+                    final String persistentUrl = mediaUpload.uploadMedia(url.openStream(), getFileName(sourceUrl));
+
+                    final Metadata metadata = newMessageMetadata(messageId, getMessageKey(sourceUrl), persistentUrl);
+                    storeMetadata(metadata);
+                }
+            } catch (MalformedURLException exception) {
+                // If it's not a URL, this is an error on the source side
+                log.warn("Source data url field is not a URL", exception);
+            } catch (Exception exception) {
+                log.error("Fetching message source content failed {}", messageMediaRequest);
+            }
+        }
+    }
+
+    private void storeMetadata(Metadata metadata) {
+        final String metadataKey = getId(metadata).toString();
+        try {
+            producer.send(new ProducerRecord<>(applicationCommunicationMetadata, metadataKey, metadata)).get();
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
```

Comment:
based on how we use the method (and that it's private) I'd probably add the exception to the method definition here. I think we'd earn in clarity (and unless I'm missing something big... behaviour would be the same)